### ISSUE TITLE
Build errors for buildtests in parallel builds due to missing dependency

### ISSUE THREAD

**Issue by juergenchrist** [openssl/openssl#19087]
When using parallel make with enabled buildtests, we sometimes see failures like

```
In file included from test/buildtest_obj_mac.c:10:
include/openssl/obj_mac.h:12: error: unterminated #ifndef
  12 | #ifndef OPENSSL_OBJ_MAC_H
```

This happens because `test/buildtest_obj_mac.c` does not have an explicit dependency on `include/openssl/obj_mac.h` which is regenerated in parallel due to the rules added in commit d040a1b9a028e89f0a33b36fb99f0151d2fdd4c3.  The rebuild path in the makefile is:

```
build_sw
->
_build_sw
->
build_libs_nodep
->
libcrypto.pc
->
libcrypto.so
->
libcrypto.so.3
->
libcrypto.ld
->
util/libcrypto.num
->
include/openssl/obj_mac.h (via CRYPTOHEADERS)
```

Since there is no dependency between this chain and the compilation of `build_test_obj_mac.c` into `test/buildtest_c_obj_mac-bin-buildtest_obj_mac.o`, they can run in parallel.  In rare cases, this leads to the situation where regeneration of obj_mac.h is still taking place while compilation reads that (partial) file causing above mentioned error.

I am not sure how to fix this issue.  Once solution would be to add an explicit dependency between the compilation step and the generated header.

Note that this problem does not occur always since parallel builds are non-deterministic.

**Comment by levitte** [comment]
An explicit dependency might not help.  With something like this:

``` make
all: 1.c 2.c
1.c: something.h
2.c: something.h
something.h:
	echo '#ifndef SOMETHING_H' > something.h
	echo '# define SOMETHING 1' > something.h
	echo '#endif' > something.h
```

... there's no saying in what order things will happen if you do `make -j8`, and in what state `something.h` will be when 1.c or 2.c is compiled, 'cause you might have `something.h` re-created twice, all depending on what state it's in when each subprocess of `make` checks it.

This means that d040a1b9a028e89f0a33b36fb99f0151d2fdd4c3, for all its good intentions, wasn't a good idea after all, and should probably be reverted.

**Comment by DDvO** [comment]
Sorry to hear that there is an issue with parallel `make` and a missing dependency on `include/openssl/obj_mac.h`.

@levitte, I presume that in your above example the 2nd and 3rd `>` should actually be `>>`.
Anyway, good point that atomicity of generating `include/openssl/obj_mac.h` is a complication here.
I suppose this can be solved by doing the stage-wise generation on a tmp file (that is local per `make` thread) and then renaming or copying it to `include/openssl/obj_mac.h`.

**Comment by levitte** [comment]
> @levitte, I presume that in your above example the 2nd and 3rd `>` should actually be `>>`.

Correct.  Fixed.

**Comment by DDvO** [comment]
Can you give a hint how to get such a local tmp file (name)?
Then I can give the above idea a try.

**Comment by ifranzki** [comment]
Something like this might do it:
```
TMPFILE := $(shell mktemp)

all: 1.c 2.c
1.c: something.h
2.c: something.h
something.h:
	echo '#ifndef SOMETHING_H' > $(TMPFILE)
	echo '# define SOMETHING 1' >> $(TMPFILE)
	echo '#endif' >> $(TMPFILE)
	mv $(TMPFILE) $@
```

**Comment by DDvO** [comment]
Thx,`$(shell mktemp)` looks good.
Yet I fear the `make` variable `TMPFILE` will be not local to a `make` thread.
So instead I'll use a shell variable set in the `something.h` rule.

**Comment by levitte** [comment]
Careful.  Not all makes are GNU make.

**Comment by levitte** [comment]
> Anyway, good point that atomicity of generating `include/openssl/obj_mac.h` is a complication here. I suppose this can be solved by doing the stage-wise generation on a tmp file (that is local per `make` thread) and then renaming or copying it to `include/openssl/obj_mac.h`.

That might work.  But, *rename* the file to its final destination.  Don't copy, that's just recreating the problem.

**Comment by ifranzki** [comment]
Well, be careful, make starts a new shell process for each recipe line. So variables you set on one line are lost on the next line. So you would need to join the recipe lines using \ to have the use the same shell:
```
foo: bar baz
    line1; \
    line2; \
    line3
```

**Comment by ifranzki** [comment]
Another option would be to use a lockfile (untested but inspired by https://www.cmcrossroads.com/article/pitfalls-and-benefits-gnu-make-parallelization):

```
LOCK_FILE := lock.me

all: 1.c 2.c
1.c: something.h
2.c: something.h
something.h:
	lockfile $(LOCK_FILE)
	echo '#ifndef SOMETHING_H' > something.h
	echo '# define SOMETHING 1' >> something.h
	echo '#endif' >> something.h
	rm -f $(LOCK_FILE)
```

**Comment by DDvO** [comment]
This looks more elegant, will try that.
Should this work with all flavors of `make`?

**Comment by juergenchrist** [comment]
How should that work?  The problem is not two concurrent executions of the `something.h` rule, but one execution of that rule and another rule that reads the file `something.h`.  This second rule is not locked at all and cannot be since it is implicitly created by make-depend.

**Comment by DDvO** [comment]
Good point.
And it turns out that `lockfile` must be some (shell) command that is not generally available, so anyway cannot use that.

**Comment by ifranzki** [comment]
`lockfile` seems to be in the procmail package, so this would be a requirement then....

But as @juergenchrist mentioned, it will only solve the problem that the header file is created atomically, but not that the compile rules have no dependency on the header file being generated at all.

**Comment by juergenchrist** [comment]
Sorry, should have seen that earlier.  @levitte gave a bad example above.  This example works perfectly (otherwise `make` would be broken).  Parallelization only takes place if the targets are independent.  Unfortunately this is the case for a lot of c files whose dependencies are automatically generated.  By that, if the depend on `obj_mac.h` that dependency will only be seen after dependencies are generated (by-product of compilation) and the regeneration of `obj_mac.h` can run in parallel.  I think the correct solution is to make sure `obj_mac.h` is up to date before starting compilation.  Otherwise there will always be corner cases where an explicit dependency is missing.

**Comment by ifranzki** [comment]
Does `include/openssl/obj_mac.h` always need to be generated before it can be used, or is it available in the source tree? If it would be available in the source tree, then making sure it is atomically re-generated might help anyway. It would just make sure that `include/openssl/obj_mac.h` always exists and is complete, either theoriginal version from the source tree, or freshly generated.

**Comment by juergenchrist** [comment]
`include/openssl/obj_mac.h` exists in the source tree.  But regeneration while the object is used does not sound like a good idea.  What if regeneration actually changes the file?  I am still not 100% sure why in our internal CI systems sometimes we recreate that file and sometimes not.  But if it is regenerated, it should be regenerated before it is ever used.

**Comment by ifranzki** [comment]
>  I am still not 100% sure why in our internal CI systems sometimes we recreate that file and sometimes not.

I guess before commit https://github.com/openssl/openssl/commit/d040a1b9a028e89f0a33b36fb99f0151d2fdd4c3 it was not re-generated at all, otherwise we should have seen the build failure before already from time to time.

> But if it is regenerated, it should be regenerated before it is ever used.

agreed.

**Comment by levitte** [comment]
> `include/openssl/obj_mac.h` exists in the source tree.

Correct, it gets generated with a manual `make update`.  The idea with #15224 was to ensure that only those files that really need regenerating are regenerated.  I actually forgot this, because we do have other header files that are generated when building, i.e. I got confused.

The thing that makes this suddenly problematic is that dependency generation, which happens when object files are rebuilt, and gets picked up on the next make run, now trigger regeneration (because those header files are now explicit targets with actions), where it previously was just a timestamp thing. 

> What if regeneration actually changes the file?

In that case, something else is badly broken.  The scripts that do generate them are designed to make this exactly reproducible.

> I am still not 100% sure why in our internal CI systems sometimes we recreate that file and sometimes not. But if it is regenerated, it should be regenerated before it is ever used.

**Comment by DDvO** [comment]
So would an order-only dependency help here?

**Comment by juergenchrist** [comment]
I think the easiest solution is to move `include/openssl/obj_mac.h` to the `GENERATED_MANDATORY` list and leave the rest as is.  Of course, the same has to be done with every header file that could be regenerated via a file rule.

**Comment by DDvO** [comment]
Meanwhile I tried the atomic generation like this:
```
$(SRCDIR)/include/openssl/obj_mac.h: $(SRCDIR)/crypto/objects/objects.pl \
                                     $(SRCDIR)/crypto/objects/obj_dat.pl \
                                     $(SRCDIR)/crypto/objects/obj_mac.num \
                                     $(SRCDIR)/crypto/objects/obj_compat.h
	# Using TMP_obj_mac_h for atomic generation
	( export TMP_obj_mac_h=`mktemp`; \
	  cd $(SRCDIR); $(PERL) crypto/objects/objects.pl \
				crypto/objects/objects.txt \
				crypto/objects/obj_mac.num \
				> $TMP_obj_mac_h && \
			$(PERL) crypto/objects/obj_dat.pl \
				$TMP_obj_mac_h \
				> crypto/objects/obj_dat.h && \
			sed -e '1,8d' crypto/objects/obj_compat.h \
				>> $TMP_obj_mac_h; \
			mv $TMP_obj_mac_h include/openssl/obj_mac.h )
```
It appears to work, but now I'm uncertain if is useful/relevant here.

**Comment by ifranzki** [comment]
> I usually simply append a .tmp to the temporary file.

You need a unique temporary file per make thread. Just adding `.tmp` will produce the same temp file for all threads, and thus you have the same concurrency problem again.....

**Comment by juergenchrist** [comment]
> > I usually simply append a .tmp to the temporary file.
> 
> You need a unique temporary file per make thread. Just adding `.tmp` will produce the same temp file for all threads, and thus you have the same concurrency problem again.....

No.  You refer to a reentrency problem that simply does not exist.  Make does not execute the same rule concurrently (that would be a bug in make or a weird setup with recursive make).  The problem is that the rule generating the file and multiple rules using the file (implicitly via gcc) run concurrently.  But if you really want a unique file name, simply use `$$` (appropriately quoted since the shell should evaluate this) to get the PID of the shell.  That should be unique enough.

**Comment by ifranzki** [comment]
>  Make does not execute the same rule concurrently (that would be a bug in make or a weird setup with recursive make). 

I have seen this happening with another project, but yes, this was a weird recursive make situation.
So you are right, this should not be the problem here.

I wonder why it is re-generating obj_mac.h at all. It is available in the source tree, so there is no need to regenerate it. At least not in the sequence we use in our internal CI: clone from git, configure, make. So that header file should be current, no?

**Comment by juergenchrist** [comment]
> clone from git, configure, make. So that header file should be current, no?

That is exactly the point I do not understand.  We run the same sequence with the same repository on different machines.  Sometimes we do not regenerate, sometimes we do.  My current (and only) explanation is that some weird time stamp problem is occurring.  Nevertheless, the build system should be able to handle this gracefully.

**Comment by levitte** [comment]
when you update a worktree with git, the files that do get changed get the timestamp of the update, none other.  That can be a reason to see "spurious" regenerations like this.  The problematic header depends on four other files, of which some are also generated.  It's enough that one of those gets an updated timestamp.

**Comment by levitte** [comment]
> ... But I guess the solution to add these files to `GENERATED_MANDATORY` (and do that for all files that have file rules and are generated) would be the best solution.

That is, indeed a solution.  In that case, they will end up in the build tree rather than the source tree (when you build in a separate build directory tree)...  and would get removed from the source tree, as well as from `make update`.  I think that would be fine.

**Comment by levitte** [comment]
> > ... But I guess the solution to add these files to `GENERATED_MANDATORY` (and do that for all files that have file rules and are generated) would be the best solution.
> 
> That is, indeed a solution. In that case, they will end up in the build tree rather than the source tree (when you build in a separate build directory tree)... and would get removed from the source tree, as well as from `make update`. I think that would be fine.

UPDATE: it's easy to ensure they end up in that variable.  That's what these lines do:

https://github.com/openssl/openssl/blob/d9aca2dd9b56dbfa6a0566cc3ad8b7c713ac61b2/build.info#L17-L43

**Comment by DDvO** [comment]
So a viable solution would be to simply add 
```
crypto/objects/obj_mac.num
include/openssl/obj_mac.h
crypto/objects/obj_a
crypto/objects/obj_xref.h
```
to that `DEPEND` array?

**Comment by levitte** [comment]
> So a viable solution would be to simply add
> 
> ```
> crypto/objects/obj_mac.num
> include/openssl/obj_mac.h
> crypto/objects/obj_a
> crypto/objects/obj_xref.h
> ```
> 
> to that `DEPEND` array?

Yes, and remove them as dependencies for `update:` in Configurations/unix.Makefile.tmpl

**Comment by DDvO** [comment]
> > So a viable solution would be to simply add
> > ```
> > crypto/objects/obj_mac.num
> > include/openssl/obj_mac.h
> > crypto/objects/obj_dat.h
> > crypto/objects/obj_xref.h
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > to that `DEPEND` array?
> 
> Yes, and remove them as dependencies for `update:` in Configurations/unix.Makefile.tmpl

Done in #19178

**Comment by DDvO** [comment]
For the records: new fix is in #19179.
